### ISSUE TITLE
Elemental Craft not working on Crafting Station

### ISSUE THREAD

**Issue by Diversion98** [SlimeKnights/TinkersConstruct#4446]
**Issue description:**
When I progress to a new stage, mods will unlock thanks to gamestages.
We only have trouble with the TC crafting station.

the workbench from vanilla and the crafting station from the craftion station mod works perfectly.

if you command the following code in the script with //, it will work again.
mods.recipestages.Recipes.setRecipeStageByMod("stage1", "minecraft");

**How to reproduce:**
1. Go to config -> gamestages and edit known_stages.json to add ["stage1"]

2. Make a Zenscipt file and include this code in it:
   mods.recipestages.Recipes.setRecipeStageByMod("stage1", "minecraft");

3. Make a new world and give yourself stage 1 with the command "/gamestage add *playername* stage1"

4. Try to craft something.

**Versions:**
Forge 36.1.16
Bookshelf 10.0.7
CraftTweaker 7.1.0.278
Game Stages 7.1.2
Just Enough Items 7.6.4.90
Mantle 1.6.92
Recipe Stages 2.0.0.14
Tinkers Construct 3.0.3.144

**Confirm below that this issue is not covered on the roadmap or "Whats New?"**
confirmed

**Can it be reproduced with *just* Tinkers Construct? If not, list the other mods *required* to reproduce the issue.**
Bookshelf 10.0.7
CraftTweaker 7.1.0.278
Game Stages 7.1.2
Just Enough Items 7.6.4.90
Mantle 1.6.92
Recipe Stages 2.0.0.14

**Comment by KnightMiner** [comment]
Have you tested with only the two mods? I see no reason a single mods recipes would not work unless they use a custom crafting table recipe serializer

**Comment by Sirttas** [comment]
I tested it with only EC and TiC (and their dependancies) and everything is working fine, I'm closing the issue on elementalcraft github.

**Comment by KnightMiner** [comment]
I don't follow the logic that you leave this open but close that one when the issue does not happen with just the two mods.

Find a minimal set of mods to reproduce the issue. Odds are your performance enhancing mods are to blame, but it could be another.

**Comment by Sirttas** [comment]
> I don't follow the logic that you leave this open but close that one when the issue does not happen with just the two mods.

I an not the creator of the issue but the owner of the other repo (author of elemental craft). I was just telling you that I closed the issue on my repo because I think it is an issue for the pack creator for a few reason:
- worked fine with only EC and TiC, it also works fine in ATM6
- I the pack the issue originate from crash on worldgen
- I read the code of Tic on this part and I don't see how it cannot work.

TLDR: I'm telling you I think you can close this issue.

**Comment by KnightMiner** [comment]
Oh sorry, I mistook you as the poster. Yeah, I guess there is not much point in keeping this open without a more minimal test case. So if the original poster can find an exact list of mods causing this, we can look into it again.

**Comment by Diversion98** [comment]
Okay I reduced it to 9 mods now.

I updated the original post

**Comment by Sirttas** [comment]
I see there is craft tweaker, what are the scripts you used ?

**Comment by Diversion98** [comment]
look original post
Only that line of code

And in the full pack, i have 100+ lines of those codes for other mods and those work just fine

**Comment by Diversion98** [comment]
eddited the original post again

Cyclic has problems as well

**Comment by KnightMiner** [comment]
So, you added a mod to a game stage, then wondered why you could not craft recipes from that mod, but did not think game stages was worth mentioning? I assume this issue has nothing to do with any mod you locked behind game stages. It's a game stages incompatibility

While this is not a known bug, this is expected that the crafting station would not work with game stages. Please open a new issue for game stages incompatibility

----
Also, you do not need 9 mods to reproduce this bug. You can do it with fewer. 9 is a more manageable number but you might as well get the final minimal list. Most bugs don't need more than 3.

**Comment by Diversion98** [comment]
I will open a new issue with gamestages then.
Also I do need at least 9 mods because they are all dependecies.

recipestages needs gamestages which needs crafttweaker, bookshelf and JEI. (without JEI it crashes)
TC needs mantle
elemental craft needs datapack anvil.

and I came to this github because every other crafting thing works except TC.
next time be more polite, i dont build mods so i dont know how u make your mods.

**Comment by KnightMiner** [comment]
That is a lot of dependencies, quite surprising to me. Anyways, as I said we can leave out elemental craft.

Sorry, my answer was a bit rude. I do expect people who are configuring modpacks to talk about how they configured them before jumping to "its a mod bug", but there was a nicer way to say that.

**Comment by Diversion98** [comment]
I accept your apology :)

I didnt knew that just a simple craft could mess with a mod cause i wasnt configuring a recipe or anything so yeah my bad.
Thanks for letting me know tho, first modpack that im making.

1 question tho, gamestages or recipestages to make an issue? :/

**Comment by KnightMiner** [comment]
Whichever mod locks the crafting recipes. Probably recipe stages?

After reporting it to them, make an issue here and link the two issues so I can keep track of this

**Comment by jaredlll08** [comment]
So I can reproduce *something* but I don't know if it is what is described in this issue.

So firstly I'm using the mods defined in the tinkers build.gradle, so TOP, JEI, Mantle.

I've added CraftTweaker, GameStages, BookShelf and RecipeStages.

I'm using 
```zenscript
mods.recipestages.Recipes.setRecipeStageByMod("stage1", "minecraft");
```
as the test script for gamestages.

So the result is shown in the output slot of the Crafting Station, but you can't actually take the result, I've recorded how it looks
![tcon](https://user-images.githubusercontent.com/5553605/117856789-5a97de80-b28c-11eb-9541-2700e1f8a4dd.gif)

I didn't see anything glaringly wrong with the code, and `slimeknights.tconstruct.tables.tileentity.table.CraftingStationTileEntity#onCraft` returns the correct `result`, so I'm not sure where the issue lies.

**Comment by Diversion98** [comment]
Yes thats the issue.

And if u shift click the result, u will get it.
But if you then open the crafting station again, the item will be gone in your inventory and back into the crafting station.

U can even duplicate the item:

1. Shift click result
2. In your inventory, place it in any other slot (doesnt matter which one)
3. Open crafting station again and it will give you the same recipe with the required ingredients and you can repeat it

**Comment by jaredlll08** [comment]
I looked into it more, the issue is that Tinkers is passing in `null`, meaning that recipes aren't able to get the Container that is doing the craft, thus they can't get the player.

The reason it shows something an output is because we return the result on the client if the player is allowed to craft it, but the server is still the final say, which since it can't get a player, it says no.

This will be an issue for any mod that has a custom crafting table recipe that requires a player, I am unaware of ones currently in 1.16.5, but in 1.12 Extra Utilities is doing the same thing that I am for their XP required recipes.
